pkg/webhook/cspc_operations.go

```diff
@@ -509,3 +468,213 @@ func GetNewBDFromRaidGroups(newRG, oldRG *cstor.RaidGroup) map[string]string {
+		// added for expansion
+		if !isRaidGroupExist {
+			expandedRgs = append(expandedRgs, newRg)
+		}
+	}
+	return expandedRgs
+}
+
+// validateNewBDs returns nil if new BDs are valid for expansion or else it will
+// return error if occurs
+func (pOps *PoolOperations) validateNewBDs(newBDs []string, cspc *cstor.CStorPoolCluster) error {
+	for _, bd := range newBDs {
+		bdc, err := pOps.GetBDCOfBD(bd)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get claim of block device %s", bd)
+		}
+		// The incoming BD should be a valid entry if
+		// 1. The BD does not have a BDC.
+		// 2. The BD has a BDC with the current CSPC label and there is no successor of this BD
+		//    present in the CSPC.
+		if !pOps.IsBDValid(bd, bdc, cspc) {
+			return errors.Errorf("can not use blockdevice %s validation failed", bd)
+		}
+	}
+	return nil
+}
+
+// validatePoolExpansion will validate only expanded raid groups or new blockdevices(
+// in stripe only block devices are added). Following are the validations:
+// 1. New blockdevice shouldn't be claimed by any other CSPC (or) third party.
+// 2. New blockdevice shouldn't be the replacing blockdevice.
+func (pOps *PoolOperations) validatePoolExpansion(
+	newPoolSpec *cstor.PoolSpec, commonRaidGroups map[string]*raidGroups) error {
+	var bds []string
+	for rgType, rgs := range commonRaidGroups {
+		if rgs.rgType == string(cstor.PoolStriped) {
+			bds = append(bds, getNewBDsFromStripeSpec(rgs.oldRaidGroups[0],
+				rgs.newRaidGroups[0])...)
+		} else {
+			if rgType == "data" {
+				newRgs := getExpandedRaidGroups(newPoolSpec.DataRaidGroups, rgs.oldRaidGroups)
+				bds = getBDsFromRaidGroups(newRgs)
+			}
+			if rgType == "writeCache" {
+				newRgs := getExpandedRaidGroups(newPoolSpec.WriteCacheRaidGroups, rgs.oldRaidGroups)
+				bds = getBDsFromRaidGroups(newRgs)
+			}
+		}
+	}
+	err := pOps.validateNewBDs(bds, pOps.NewCSPC)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// getIndexedCommonRaidGroups returns raidGroups that contains index matching of
+// oldRaidGroups and newRaidGroups. If oldRaidGroup doesn't exist on
+// newRaidGroup then return error
+func getIndexedCommonRaidGroups(oldPoolSpec,
+	newPoolSpec *cstor.PoolSpec) (map[string]*raidGroups, error) {
+	rgs := map[string]*raidGroups{
+		"data": &raidGroups{
+			oldRaidGroups: []cstor.RaidGroup{},
+			newRaidGroups: []cstor.RaidGroup{},
+			rgType:        oldPoolSpec.PoolConfig.DataRaidGroupType,
+		},
+		"writeCache": &raidGroups{
+			oldRaidGroups: []cstor.RaidGroup{},
+			newRaidGroups: []cstor.RaidGroup{},
+			rgType:        oldPoolSpec.PoolConfig.WriteCacheGroupType,
+		},
+	}
+	// build raidGroups by identifying common raidGroups in old and new
+	for _, oldRg := range oldPoolSpec.DataRaidGroups {
+		isRaidGroupExist := false
+		for _, newRg := range newPoolSpec.DataRaidGroups {
+			if IsRaidGroupCommon(oldRg, newRg) {
+				isRaidGroupExist = true
+				rgs["data"].oldRaidGroups = append(rgs["data"].oldRaidGroups, oldRg)
```

Comment:
nit: IMO good to maintain global variable for the type of the disk instead of literals